### PR TITLE
Add lock to ensure that alembic commands are not run concurrently

### PR DESCRIPTION
**Pull request by madkinsz**
Co-authored with @ahuang11 
Adds locks around alembic commands.

Alembic commands rely on mutation of shared global state. Running multiple commands concurrently can result in corrupted configurations. To prevent this, we add acquisition of a shared lock to all alembic commands.

See:
-  #7720
- https://github.com/PrefectHQ/prefect-dask/pull/50
- https://github.com/PrefectHQ/prefect-dask/issues/55

Note: In https://github.com/PrefectHQ/prefect/pull/7789/commits/b43e04eb3f616e8e7ba60e999b26b046c99ada25 I attempted to use async locks in the database interface instead. This resulted in deadlocks when alembic was used from multiple threads via separate event loops. Async locks are not thread safe and this approach is safer.
 

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

**Comment by madkinsz** [comment]
@ahuang11 is going to take ownership of this as it is related to some Dask work.

I have not checked if it resolves compatibility with Flask, cc @meystingray

**Review by jakekaplan — APPROVED** [review]
Nice!

File: tests/orion/database/test_alembic_commands.py [hunk 1]
```diff
@@ -69,3 +71,15 @@ def test_alembic_stamp(self, mocked):
         _, kwargs = mocked.call_args
         assert mocked.call_count == 1
         assert kwargs["revision"] == "abcdef"
+
+    async def test_concurrent_upgrade(self):
+        jobs = [run_sync_in_worker_thread(alembic_upgrade) for _ in range(0, 10)]
+        await asyncio.gather(*jobs)
+
+    @pytest.mark.flaky(max_runs=3)
```
Comment (madkinsz) [inline tests/orion/database/test_alembic_commands.py:79]: Does this need to be flaky?

Comment (ahuang11) [reply tests/orion/database/test_alembic_commands.py]: I had it `range(0, 3)` and it timed out; let me see if 2 times out.

**Pull request merged**